File: solutions/csharp/food-chain/1/FoodChain.cs
#nullable enable

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

public static class FoodChain
{
	private static readonly Dictionary<int, Animal> _animalByNumber = new()
	{
		{ 1, new Animal("fly") },
		{ 2, new Animal("spider", "It wriggled and jiggled and tickled inside her.", true) },
		{ 3, new Animal("bird", "How absurd to swallow a bird!") },
		{ 4, new Animal("cat", "Imagine that, to swallow a cat!") },
		{ 5, new Animal("dog", "What a hog, to swallow a dog!") },
		{ 6, new Animal("goat", "Just opened her throat and swallowed a goat!") },
		{ 7, new Animal("cow", "I don't know how she swallowed a cow!") },
		{ 8, new Animal("horse", "She's dead, of course!") },
	};
	private static readonly int _lastAnimal = _animalByNumber.Keys.Max();

	public static string Recite(int verseNumber)
	{
		if (!_animalByNumber.TryGetValue(verseNumber, out Animal animal))
		{
			throw new ArgumentOutOfRangeException(nameof(verseNumber), $"Neither animal nor verse exists for {verseNumber}");
		}

		StringBuilder stringBuilder = new($"I know an old lady who swallowed {animal.IndefiniteForm}.");
		if (!string.IsNullOrEmpty(animal.Remark))
		{
			stringBuilder = stringBuilder.Append($"\n{animal.Remark}");
		}

		StringBuilder refrain = CreateRefrain(verseNumber);
		return stringBuilder.Append(refrain)
			.ToString();

	}

	private static StringBuilder CreateRefrain(int verseNumber)
	{
		StringBuilder stringBuilder = new();
		if (verseNumber != _lastAnimal)
		{
			stringBuilder = stringBuilder.Append('\n');

			for (int animalNumber = verseNumber; animalNumber > 1; animalNumber--)
			{
				string swallowLine = CreateSwallowLine(animalNumber);
				stringBuilder = stringBuilder.Append(swallowLine);
			}

			stringBuilder = stringBuilder.Append("I don't know why she swallowed the fly. Perhaps she'll die.");
		}

		return stringBuilder;
	}

	private static string CreateSwallowLine(int animalNumber)
	{
		Animal currentAnimal = _animalByNumber[animalNumber];
		Animal previousAnimal = _animalByNumber[animalNumber - 1];
		string lineEnd = previousAnimal.HasExtendVerse
			? $" {previousAnimal.VersExtension}"
			: ".";
		return $"She swallowed the {currentAnimal.Name} to catch the {previousAnimal.Name}{lineEnd}\n";
	}

	public static string Recite(int startVerse, int endVerse)
	{
		List<string> verses = Enumerable.Range(startVerse, endVerse)
			.Select(x => Recite(x))
			.ToList();
		return string.Join("\n\n", verses.ToArray());
	}
}

public record struct Animal(string Name, string? Remark = null, bool HasExtendVerse = false)
{
	public string IndefiniteForm =>
		Name.PrefixArticle();

	public string VersExtension =>
		(!HasExtendVerse || string.IsNullOrWhiteSpace(Remark))
			? throw new InvalidOperationException($"{Name} does not have extended verse")
			: Remark.Replace("It", "that");
}

public static class StringExtensions
{
	private static readonly HashSet<char> vowels = new()
	{
		'a', 'e', 'i', 'o', 'u'
	};

	public static string PrefixArticle(this string text) =>
		string.IsNullOrWhiteSpace(text)
			? throw new ArgumentException($"'{nameof(text)}' cannot be null or whitespace.", nameof(text))
			: StartsWithVowel(text)
				? $"an {text}"
				: $"a {text}";

	private static bool StartsWithVowel(string text)
	{
		char firstCharacter = char.ToLowerInvariant(text[0]);
		return vowels.Contains(firstCharacter);
	}
}

File: solutions/java/food-chain/1/src/main/java/FoodChain.java
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public final class FoodChain {
    private static final Map<Integer, Animal> ANIMAL_BY_NUMBER = Map.of(
            1, Animal.FLY,
            2, Animal.SPIDER,
            3, Animal.BIRD,
            4, Animal.CAT,
            5, Animal.DOG,
            6, Animal.GOAT,
            7, Animal.COW,
            8, Animal.HORSE
    );

    private static final int LAST_ANIMAL = Collections.max(ANIMAL_BY_NUMBER.keySet());

    public String verse(int verseNumber) {
        Animal animal = Optional.ofNullable(ANIMAL_BY_NUMBER.get(verseNumber))
                                .orElseThrow(() -> new IllegalArgumentException(
                                        "Neither animal nor verse exists for " + verseNumber));

        return this.createVerse(verseNumber, animal);
    }

    private String createVerse(final int verseNumber, final Animal animal) {
        final String remark = animal.getRemark();
        String verse = "I know an old lady who swallowed " + animal.indefiniteForm() + ".";
        if (remark != null && !remark.isBlank()) {
            verse += "\n" + remark;
        }
        verse += this.createRefrain(verseNumber);
        return verse;
    }

    private String createRefrain(int verseNumber) {
        if (verseNumber == LAST_ANIMAL) {
            return "";
        }

        String firstNewline = "\n";

        String lines = IntStream.iterate(verseNumber, n -> n > 1, n -> n - 1)
                                .mapToObj(this::createSwallowLine)
                                .collect(Collectors.joining());

        return firstNewline + lines + "I don't know why she swallowed the fly. Perhaps she'll die.";
    }

    private String createSwallowLine(int animalNumber) {
        Animal current = ANIMAL_BY_NUMBER.get(animalNumber);
        Animal previous = ANIMAL_BY_NUMBER.get(animalNumber - 1);

        String lineEnd = previous.hasExtendVerse()
                         ? " " + previous.verseExtension()
                         : ".";

        return "She swallowed the " + current.getName() +
               " to catch the " + previous.getName() + lineEnd + "\n";
    }

    public String verses(int startVerse, int endVerse) {
        return IntStream.rangeClosed(startVerse, endVerse)
                        .mapToObj(this::verse)
                        .collect(Collectors.joining("\n\n"));
    }
}
